### PR TITLE
Improve memcpy

### PR DESCRIPTION
**Pull request by agurfinkel**
Support memcpy of unaligned regions

### CODE REVIEW

File: lib/seahorn/BvOpSem2MemRepr.cc [hunk 1]
```diff
@@ -358,12 +370,50 @@ OpSemMemLambdaRepr::MemCpy(PtrTy dPtr, PtrTy sPtr, unsigned len,
...
+        Expr lastWordValSrc =
+            op::bind::fapp(srcMem.toExpr(), readPtrInSrc.toExpr());
+
+        // -- compute the last word by taking chunks of source and destination
+        // -- words. source word comes first
+        unsigned wordSzInBits = wordSzInBytes * 8;
+        unsigned remainderBits = remainderBytes * 8;
+        auto &alu = m_ctx.alu();
+        Expr srcChunk =
+            alu.Extract({lastWordValSrc, wordSzInBits}, 0, remainderBits - 1);
```
Comment (priyasiddharth) [inline lib/seahorn/BvOpSem2MemRepr.cc:398]: Please confirm that `end` is inclusive

i.e. if `remainderBits = 4` then bits 0, 1, 2, 3 will be extracted.

File: lib/seahorn/BvOpSem2MemRepr.cc [hunk 2]
```diff
@@ -347,8 +347,20 @@ OpSemMemLambdaRepr::MemCpy(PtrTy dPtr, PtrTy sPtr, unsigned len,
...
+      unsigned lastByteToCopy;
+      unsigned remainderBytes;
+      if (m_memManager.isIgnoreAlignment()) {
+        // if alignment is ignored, we treat it as alignment of 1
+        lastByteToCopy = len - 1;
+        remainderBytes = 0;
+      } else {
+        unsigned wordsToCopy = (len / wordSzInBytes);
+        // -- -1 because ptrInRangeCheck is inclusive
+        lastByteToCopy = (wordsToCopy - 1) * wordSzInBytes;
```
Comment (priyasiddharth) [inline lib/seahorn/BvOpSem2MemRepr.cc:359]: This is really `lastAlignedBytePosToCopy` isn't it? Do you think that is a better name?

**Review by priyasiddharth — COMMENTED** [review]
Minor questions for my understanding.

File: lib/seahorn/BvOpSem2MemRepr.cc (hunk 1 above)
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2MemRepr.cc:398]: yes. definitely inclusive. Forgot `-1` at first and z3 complained with type error.

File: lib/seahorn/BvOpSem2MemRepr.cc (hunk 2 above)
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2MemRepr.cc:359]: done. will commit soon.

**Comment by agurfinkel** [comment]
@priyasiddharth : good to merge?
@danblitzhou : ? (this is related to what you doing, so would be great if you can understand this code)

**Pull request merged**